lib/wibox/widget/separator.lua

```diff
@@ -0,0 +1,188 @@
+---------------------------------------------------------------------------
+local beautiful = require( "beautiful"         )
+local base      = require( "wibox.widget.base" )
+local color     = require( "gears.color"       )
+local gtable    = require( "gears.table"       )
+
+local separator = {}
+
+--- The separator's orientation.
+--
+-- Valid values are:
+--
+-- * *vertical*: From top to bottom (default)
+-- * *horizontal*: From left to right
+--
+--@DOC_wibox_widget_separator_orientation_EXAMPLE@
+--
+-- @property orientation
+-- @param string
+
+--- The separator's thickness.
+--
+-- This is used by the default line separator, but ignored when a shape is used.
+--
+-- @property thickness
+-- @param number
+
+--- The separator's reflection.
+--
+-- When a shape or a custom drawing function is used, the default shape assumes
+-- this to go from left to right.
+-- When this property is set to anything else, the shape will be rotated on one
+-- axis or another.
+--
+-- This is the same as `wibox.container.mirror`, but need to exist on the widget
+-- level because it affects other properties such as the layout clipping.
+--
+-- @property reflection
+-- @tparam table reflection
+-- @tparam[opt=false] boolean reflection.horizontal
+-- @tparam[opt=false] boolean reflection.vertical
+
+--- The separator's shape.
+--
+--@DOC_wibox_widget_separator_shape_EXAMPLE@
+--
+-- @property shape
+-- @tparam function shape A valid shape function
+-- @see gears.shape
+
+--- The relative percentage covered by the bar.
+-- @property span_ratio
+-- @tparam[opt=1] number A number between 0 and 1.
+
+--- The separator's color.
+-- @property color
+-- @param string
+-- @see gears.color
+
+--- The separator's border color.
+--
+--@DOC_wibox_widget_separator_border_color_EXAMPLE@
+--
+-- @property border_color
+-- @param string
+-- @see gears.color
+
+--- The separator's border width.
+-- @property border_width
+-- @param number
+
+local function draw_shape(self, _, cr, width, height, shape)
+    local bw = self._private.border_width or 0
+
+    cr:translate(bw/2, bw/2)
+
+    shape(cr, width-bw, height-bw)
+
+    if bw == 0 then
+        cr:fill()
+    elseif self._private.border_color then
+        cr:fill_preserve()
+        cr:set_source(color(self._private.border_color))
+        cr:set_line_width(bw)
+        cr:stroke()
+    end
+end
+
+local function draw_line(self, _, cr, width, height)
+    local thickness = self._private.thickness or beautiful.separator_thickness or 1
+
+    local orientation = self._private.orientation or (
+        width > height and "horizontal" or "vertical"
+    )
+
+    local span_ratio = self._private.span_ratio or 1
+
+    if orientation == "horizontal" then
+        local w = width*span_ratio
+        cr:rectangle((width-w)/2, height/2 - thickness/2, w, thickness)
+    else
+        local h = height*span_ratio
+        cr:rectangle(width/2 - thickness/2, (height-h)/2, thickness, h)
+    end
+
+    cr:fill()
+end
```

Comment:
Can't this line be done as a shape? I.e. when no shape is set, this function (without the `cr:fill()` at the end) is just used instead? That would then also add a border color for the line.

Reply:
It could, but I think it would make things more complex for the same end result. At least this code is rather obvious and isn't using all the shape API indirections.